spec.md

```diff
@@ -538,6 +540,129 @@ It indicates the type of the data being stored inside of the memory.
+Each of these is a subcomponent of `v`{.firrtl} and each acts like a wire with type `UInt<8>`{.firrtl}.
+
+Next, let's look at a port with a bundle type:
+
+``` firrtl
+module Bar :
+  output io : { x : UInt<8>, flip y : UInt<8> }
+  connect io.x, add(io.y, UInt(1))
+```
+
+The bundle of port `io`{.firrtl} has type `{ x : UInt<8>, flip y : UInt<8> }`{.firrtl}, with one aligned field `x`{.firrtl} and one flipped field `y`{.firrtl}.
+In the connect expression, we read from `io.y`{.firrtl}, add 1 to it, and then assign it to `io.x`{.firrtl}.
+Both `io.x`{.firrtl} and `io.y`{.firrtl} are subcomponents of `io`{.firrtl} and both have type `UInt<8>`{.firrtl}.
+Note that while `io.x`{.firrtl} is an output port, `io.y`{.firrtl} is an input port.
+
+### Definition
+
+Every circuit component declared within a module results in a tree of **subcomponents**.
+Circuit subcomponents have both a kind and a type.
+
+We define this tree of subcomponents recursively by defining the a **direct subcomponent** relation:
+
+A circuit component with a ground type, an enumeration type, a probe type, or property type (see [@sec:types]) has no direct subcomponents.
+For example, `wire w : UInt<8>`{.firrtl} has no direct subcomponents.
+
+When a circuit component has a vector type (see [@sec:vector-types]), it has as many direct subcomponents as its length.
+Each subcomponent will have the same kind as its parent and will have the element type.
+For example, if we declare `wire v : UInt<8>[3]`{.firrtl}, it will have three direct subcomponents: `v[0]`{.firrtl}, `v[1]`{.firrtl}, and `v[2]`{.firrtl}.
+All three are wires and all three have type `UInt<8>`{.firrtl}.
+
+When a circuit component has a bundle type (see [@sec:bundle-types]), it will end up with one direct subcomponent for each field.
+The kind of the subcomponent depends on both the kind and the type of the parent:
+
+-   For nodes, wires, and registers, the kind of each direct subcomponent is the same.
+-   For output and input ports, the kind of each direct subcomponent depends on whether or not the field is flipped.
+    When the field is not flipped, the kind remains the same.
+    When the field is flipped, it will change from output to input or vice versa.
+-   For submodule instances and memories, the kind of each direct subcomponent depends on whether or not the field is flipped.
+    When the field is not flipped, the kind is an input port.
+    When the field is flipped, the kind is an output port.
+
+If the bundle is not `const`{.firrtl}, the type of each subcomponent will simply be the type of the corresponding field.
+However, if the bundle is `const`{.firrtl}, the type of each subcomponent will be the `const`{.firrtl} version of the type of the corresponding field.
+
+A circuit component is a **subcomponent** of another if there is way to get from the first component to the second through the direct subcomponent relation.
+A circuit component is trivially considered to be a subcomponent of itself.
```

Comment:
Consider packing this sentence and the previous into an enumeration as either of these can be true for this to be a subcomponent.